### PR TITLE
Rahu ifix 244

### PR DESCRIPTION
**Pull request by rahu-eGov**
Fiscal event post processor dereference.

### CODE REVIEW

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/consumer/FiscalEventPostConsumer.java [hunk 1]
```diff
@@ -22,17 +22,17 @@
...
     private ObjectMapper mapper;
-
-    @KafkaListener(topics = { "${fiscal.event.kafka.push.topic}"})
+    /**/
+    @KafkaListener(topics = {"${fiscal.event.kafka.push.topic}"})
     public void listen(final HashMap<String, Object> record, @Header(KafkaHeaders.RECEIVED_TOPIC) String topic) {
 
         try {
-            log.info("Received fiscal event post request " + record);
+            log.info("Received fiscal event post request : " + record);
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/java/org/egov/consumer/FiscalEventPostConsumer.java:30]: Remember to remove this before promoting the build to QA. 
We should not log each and every record.

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/consumer/FiscalEventPostConsumer.java [hunk 2]
```diff
@@ -22,17 +22,17 @@
...
             FiscalEventRequest fiscalEventRequest = mapper.convertValue(record, FiscalEventRequest.class);;
             deferenceService.dereference(fiscalEventRequest);
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/java/org/egov/consumer/FiscalEventPostConsumer.java:32]: Correct the Spelling Error.

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceEnrichmentService.java [hunk 3]
```diff
@@ -2,14 +2,43 @@
...
+        fiscalEventDeReferenced.setEventTime(fiscalEvent.getEventTime());
+        fiscalEventDeReferenced.setEventType(fiscalEvent.getEventType().name());
+        fiscalEventDeReferenced.setParentReferenceId(fiscalEvent.getParentReferenceId());
+        fiscalEventDeReferenced.setParentEventId(fiscalEvent.getParentEventId());
+        fiscalEventDeReferenced.setReferenceId(fiscalEvent.getReferenceId());
+        fiscalEventDeReferenced.setAttributes(fiscalEvent.getAttributes());
+        fiscalEventDeReferenced.setVersion("1.0.0");
 
-    public void enrich(FiscalEventDeReferenced fiscalEventDeReferenced, FiscalEventDeReferenced fiscalEventDeReferenced1) {
+        AuditDetails auditDetails = null;
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceEnrichmentService.java:35]: AuditDetails are already present in the consumed record from the fiscal-event-service. Why do we need it here?

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceService.java [hunk 4]
```diff
@@ -22,13 +27,124 @@
...
+    private DepartmentUtil departmentUtil;
+
     public void dereference(FiscalEventRequest fiscalEventRequest) {
         FiscalEventDeReferenced fiscalEventDeReferenced =  new FiscalEventDeReferenced();
+        dereferenceTenantId(fiscalEventRequest,fiscalEventDeReferenced);
         dereferenceCoaId(fiscalEventRequest,fiscalEventDeReferenced);
-        enricher.enrich(fiscalEventDeReferenced,fiscalEventDeReferenced);
+        dereferenceProjectId(fiscalEventRequest, fiscalEventDeReferenced);
+        enricher.enrich(fiscalEventRequest,fiscalEventDeReferenced);
         producer.push(processorConfig.getEventProcessorMongoDB(),fiscalEventDeReferenced);
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceService.java:51]: This service will return FiscalEventDeReferenced. Write the producer.push statement in the consumer.

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/util/CoaUtil.java [hunk 5]
```diff
@@ -1,10 +1,94 @@
 package org.egov.util;
 
 
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.gson.JsonArray;
+import com.jayway.jsonpath.JsonPath;
 import lombok.extern.slf4j.Slf4j;
+import net.minidev.json.JSONArray;
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/java/org/egov/util/CoaUtil.java:10]: Don't use this library. Instead use jackson JsonNode/ObjectNode/ArrayNode.

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/util/FiscalEventPostProcessorUtil.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+public class FiscalEventPostProcessorUtil {
+
+    /**
+     * Method to return auditDetails of fiscal event request
+     *
+     * @param by
+     * @param isCreate
+     * @return AuditDetails
+     */
+    public AuditDetails enrichAuditDetails(String by, AuditDetails auditDetails, Boolean isCreate) {
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/java/org/egov/util/FiscalEventPostProcessorUtil.java:19]: Not needed in post-processor.

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/util/GovernmentUtil.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            }
+        }
+        return Collections.emptyList();
+    }
+
+
+    /**
+     * @return
+     */
+    private String createSearchTenantUrl() {
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/java/org/egov/util/GovernmentUtil.java:70]: It should be createSearch**Government**Url.

File: domain-services/fiscal-event-post-processor/src/main/resources/application.properties [hunk 8]
```diff
@@ -5,9 +5,9 @@ app.timezone=UTC
 kafka.config.bootstrap_server_config=localhost:9092
 spring.kafka.consumer.value-deserializer=org.egov.tracer.kafka.deserializer.HashMapDeserializer
 spring.kafka.consumer.key-deserializer=org.apache.kafka.common.serialization.StringDeserializer
-spring.kafka.consumer.group-id=fiscal-event-post-processor
-#spring.kafka.producer.key-serializer=org.apache.kafka.common.serialization.StringSerializer
-#spring.kafka.producer.value-serializer=org.springframework.kafka.support.serializer.JsonSerializer
+spring.kafka.consumer.group-id=fiscal-event-post-event
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/resources/application.properties:8]: Name the consumer group id = project-name: "fiscal-event-post-processor"

File: domain-services/fiscal-event-post-processor/src/main/resources/application.properties [hunk 9]
```diff
@@ -39,3 +39,10 @@ ifix.master.project.host=http://localhost:8030
 ifix.master.project.context.path=/ifix-master-data
 ifix.master.project.search.path=/project/v1/_search
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/resources/application.properties:41]: Add the topic names from where we are consuming the event and to where we are pushing the event as environment variables. 
We are using following topics: 
1. "fiscal-event-request-validated" 
2. "fiscal-event-request-dereferenced"
3. "fiscal-event-line-item-flattened"

**Review by rushang7-eGov — CHANGES_REQUESTED** [review]
@pintu-eGov Make the necessary changes

**Comment by rushang7-eGov** [comment]
@pintu-eGov 
Also add 2 new topics:
1. "fiscal-event-mongodb-sink"
2. "fiscal-event-druid-sink"

To implement this, you will have to write multiple producer.push statements near to each other, and there will be duplicate data but thats okay.

1. The structure of "fiscal-event-mongodb-sink" will be same as "fiscal-event-request-dereferneced"
2. The structure of "fiscal-event-druid-sink" will be same as "fiscal-event-line-item-flattened"

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceEnrichmentService.java (hunk 3 above)
Comment (pintu-eGov) [reply domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceEnrichmentService.java:35]: yes. That make sense. will remove

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceService.java (hunk 4 above)
Comment (pintu-eGov) [reply domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceService.java:51]: As per my point of view  consumer should only consume data . service will take care of producing and other business logic on it. but anyway I'll change accordingly.

File: domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceEnrichmentService.java [hunk 10]
```diff
@@ -2,14 +2,34 @@
...
+        fiscalEventDeReferenced.setEventTime(fiscalEvent.getEventTime());
+        fiscalEventDeReferenced.setEventType(fiscalEvent.getEventType().name());
+        fiscalEventDeReferenced.setParentReferenceId(fiscalEvent.getParentReferenceId());
+        fiscalEventDeReferenced.setParentEventId(fiscalEvent.getParentEventId());
+        fiscalEventDeReferenced.setReferenceId(fiscalEvent.getReferenceId());
+        fiscalEventDeReferenced.setAttributes(fiscalEvent.getAttributes());
+        fiscalEventDeReferenced.setVersion("1.0.0");
+
+        fiscalEventDeReferenced.setAuditDetails(fiscalEvent.getAuditDetails());
+        fiscalEventDeReferenced.setId(UUID.randomUUID().toString());
```
Comment (rushang7-eGov) [inline domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceEnrichmentService.java:33]: Use uuid coming in the fiscalEvent.

Comment (pintu-eGov) [reply domain-services/fiscal-event-post-processor/src/main/java/org/egov/service/FiscalEventDereferenceEnrichmentService.java:33]: done

**Pull request merged**